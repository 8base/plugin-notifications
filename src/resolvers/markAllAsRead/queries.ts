import gql from 'graphql-tag';

export const USER_NOTIFICATIONS_QUERY = gql`
  query UserNotifications($filter: UserNotificationFilter) {
    userNotificationsList(filter: $filter) {
      items {
        id
      }
    }
  }
`;

export const UPDATE_USER_NOTIFICATION_MUTATION = gql`
  mutation UpdateUserNotification($data: UserNotificationUpdateInput!) {
    userNotificationUpdate(data: $data) {
      id
    }
  }
`;
