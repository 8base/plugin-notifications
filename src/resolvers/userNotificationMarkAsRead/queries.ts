import gql from 'graphql-tag';

export const USER_NOTIFICATION_UPDATE_MUTATION = gql`
  mutation UserNotificationUpdate($data: UserNotificationUpdateInput!) {
    userNotificationUpdate(data: $data) {
      id
    }
  }
`;
