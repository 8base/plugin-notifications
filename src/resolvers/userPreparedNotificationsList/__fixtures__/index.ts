export const TABLE_SCHEMA = {
  id: '5e3836df9564db0007a4a525',
  fields: [
    {
      name: 'id',
      displayName: 'ID',
      fieldType: 'ID',
      isSystem: true,
      relation: null,
    },
    {
      name: 'createdAt',
      displayName: 'Created At',
      fieldType: 'DATE',
      isSystem: true,
      relation: null,
    },
    {
      name: 'updatedAt',
      displayName: 'Updated At',
      fieldType: 'DATE',
      isSystem: true,
      relation: null,
    },
    {
      name: 'createdBy',
      displayName: 'Created By',
      fieldType: 'RELATION',
      isSystem: true,
      relation: {
        refTable: {
          name: 'Users',
          fields: [
            {
              id: '5e381797dcb2b80008734828',
              name: 'id',
              fieldType: 'ID',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b80008734829',
              name: 'createdAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b8000873482c',
              name: 'updatedAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e38179bdcb2b800087349b9',
              name: 'createdBy',
              fieldType: 'RELATION',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b80008734832',
              name: 'email',
              fieldType: 'TEXT',
              isSystem: true,
            },
            {
              id: '5e38179ddcb2b80008734a10',
              name: 'status',
              fieldType: 'SWITCH',
              isSystem: true,
            },
            {
              id: '5e38179ddcb2b80008734a13',
              name: 'origin',
              fieldType: 'SWITCH',
              isSystem: true,
            },
            {
              id: '5e38179bdcb2b800087349b6',
              name: 'is8base',
              fieldType: 'SWITCH',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b80008734838',
              name: 'firstName',
              fieldType: 'TEXT',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b8000873483b',
              name: 'lastName',
              fieldType: 'TEXT',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b8000873483e',
              name: 'timezone',
              fieldType: 'TEXT',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b80008734943',
              name: 'avatar',
              fieldType: 'FILE',
              isSystem: true,
            },
            {
              id: '5e381797dcb2b80008734956',
              name: 'roles',
              fieldType: 'RELATION',
              isSystem: true,
            },
            {
              id: '5e381be2dcb2b800087524ea',
              name: 'sentNotifications',
              fieldType: 'RELATION',
              isSystem: false,
            },
            {
              id: '5e3a7ceed8955e00087a635e',
              name: 'clientNotifications',
              fieldType: 'RELATION',
              isSystem: false,
            },
          ],
        },
      },
    },
    {
      name: 'notification',
      displayName: 'Notification',
      fieldType: 'RELATION',
      isSystem: false,
      relation: {
        refTable: {
          name: 'Notifications',
          fields: [
            {
              id: '5e381adf8caf0200089283ba',
              name: 'id',
              fieldType: 'ID',
              isSystem: true,
            },
            {
              id: '5e381adf8caf0200089283bb',
              name: 'createdAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e381adf8caf0200089283be',
              name: 'updatedAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e381adf8caf0200089283c4',
              name: 'createdBy',
              fieldType: 'RELATION',
              isSystem: true,
            },
            {
              id: '5e381be2dcb2b800087524e7',
              name: 'actor',
              fieldType: 'RELATION',
              isSystem: false,
            },
            {
              id: '5e381b488caf02000892ada2',
              name: 'template',
              fieldType: 'RELATION',
              isSystem: false,
            },
            {
              id: '5e38375b9564db0007a4ee0d',
              name: 'entity',
              fieldType: 'RELATION',
              isSystem: false,
            },
            {
              id: '5e3a7d3dd8955e00087ab73b',
              name: 'clientNotifications',
              fieldType: 'RELATION',
              isSystem: false,
            },
          ],
        },
      },
    },
    {
      name: 'post',
      displayName: 'Post',
      fieldType: 'RELATION',
      isSystem: false,
      relation: {
        refTable: {
          name: 'Posts',
          fields: [
            {
              id: '5e381a36d1978f00089e37f0',
              name: 'id',
              fieldType: 'ID',
              isSystem: true,
            },
            {
              id: '5e381a36d1978f00089e37f1',
              name: 'createdAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e381a36d1978f00089e37f4',
              name: 'updatedAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e381a36d1978f00089e37fa',
              name: 'createdBy',
              fieldType: 'RELATION',
              isSystem: true,
            },
            {
              id: '5e381a4e8caf020008922bf1',
              name: 'title',
              fieldType: 'TEXT',
              isSystem: false,
            },
            {
              id: '5e381a54dcb2b80008744cf3',
              name: 'message',
              fieldType: 'TEXT',
              isSystem: false,
            },
            {
              id: '5e3837209564db0007a4cbca',
              name: 'notificationEntities',
              fieldType: 'RELATION',
              isSystem: false,
            },
            {
              id: '5e3839ee3e61f90007d9ce68',
              name: 'hidden',
              fieldType: 'SWITCH',
              isSystem: false,
            },
            {
              id: '5e396a82be370200087caef2',
              name: 'comments',
              fieldType: 'RELATION',
              isSystem: false,
            },
          ],
        },
      },
    },
    {
      name: 'comment',
      displayName: 'Comment',
      fieldType: 'RELATION',
      isSystem: false,
      relation: {
        refTable: {
          name: 'Comments',
          fields: [
            {
              id: '5e396a61be370200087c8745',
              name: 'id',
              fieldType: 'ID',
              isSystem: true,
            },
            {
              id: '5e396a61be370200087c8746',
              name: 'createdAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e396a61be370200087c8749',
              name: 'updatedAt',
              fieldType: 'DATE',
              isSystem: true,
            },
            {
              id: '5e396a61be370200087c874f',
              name: 'createdBy',
              fieldType: 'RELATION',
              isSystem: true,
            },
            {
              id: '5e396a6f97c4ef0007ebcf0f',
              name: 'message',
              fieldType: 'TEXT',
              isSystem: false,
            },
            {
              id: '5e396a82be370200087caeef',
              name: 'post',
              fieldType: 'RELATION',
              isSystem: false,
            },
            {
              id: '5e396aea1f790b00082e6a7a',
              name: 'notificationEntities',
              fieldType: 'RELATION',
              isSystem: false,
            },
          ],
        },
      },
    },
  ],
};

export const NOTIFICATION_ENTITIES_SCHEMA_RESPONSE = {
  system: {
    table: TABLE_SCHEMA,
  },
};

export const NOTIFICATIONS_LIST_RESPONSE = {
  userNotificationsList: {
    items: [
      {
        notification: {
          entity: {
            post: {
              id: 'postId',
              title: 'post title',
              message: 'post message',
            },
          },
          template: {
            entityType: 'Posts',
            title: 'New post "{{ post.title }}"',
            message: '{{ actor.firstName }} {{ actor.lastName }} created a new post "{{ post.title }}"',
          },
          actor: {
            id: 'actorId',
            email: 'zouxuoz@gmail.com',
            firstName: 'Vladimir',
            lastName: 'Osipov',
          },
        },
        id: 'notficationId',
        readed: false,
      },
    ],
  },
};
