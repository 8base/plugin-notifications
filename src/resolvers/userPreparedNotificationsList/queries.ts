import gql from 'graphql-tag';

export const USER_NOTIFICATIONS_LIST_QUERY = gql`
  query UserNotifications {
    userNotificationsList {
      items {
        notification {
          template {
            entityType
            title
            message
          }
          actor {
            id
            firstName
            lastName
            email
          }
        }
        id
        readed
      }
      count
    }
  }
`;

export const TABLES_SCHEMA_QUERY = gql`
  query TablesSchema {
    system {
      tablesList {
        items {
          id
          name
          fields {
            name
            displayName
            fieldType
            isList
            isSystem
            relation {
              refTable {
                id
                name
              }
            }
          }
        }
      }
    }
  }
`;
