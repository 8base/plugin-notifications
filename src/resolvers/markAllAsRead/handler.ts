import * as R from 'ramda';

import { USER_NOTIFICATIONS_QUERY, UPDATE_USER_NOTIFICATION_MUTATION } from './queries';

type MarkAllAsReadResponse = {
  data: {
    success: boolean;
    updatedIds: string[];
    failedIds: string[];
  };
  errors?: Array<Record<string, any>>;
};

type UserNotification = {
  id: string;
};

const DEFAULT_FILTER = {
  notifier: { id: { equals: '__loggedInUserId' } },
  read: { equals: false },
};

export default async (event: any, ctx: any): Promise<MarkAllAsReadResponse> => {
  let { filter = {} } = event.data;
  filter = R.mergeDeepRight(DEFAULT_FILTER, filter);

  const updatedIds: string[] = [];
  const failedIds: string[] = [];

  const { userNotificationsList } = await ctx.api.gqlRequest(
    USER_NOTIFICATIONS_QUERY,
    { filter },
    { checkPermissions: false },
  );

  const notificationsToMark: UserNotification[] = R.pathOr([], ['items'], userNotificationsList);

  const updateNotificationPromises = notificationsToMark.map(async notification => {
    try {
      await ctx.api.gqlRequest(UPDATE_USER_NOTIFICATION_MUTATION, { data: { id: notification.id, read: true } });
      updatedIds.push(notification.id);
    } catch (e) {
      failedIds.push(notification.id);
    }
  });

  await Promise.all(updateNotificationPromises);

  return {
    data: {
      success: failedIds.length <= 0 ? true : false,
      updatedIds,
      failedIds,
    },
  };
};
