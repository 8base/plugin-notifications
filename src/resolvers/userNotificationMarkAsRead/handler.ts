import { USER_NOTIFICATION_UPDATE_MUTATION } from './queries';

type UserNotificationMarkAsReadResponse = {
  data: {
    success: boolean;
  };
};

export default async (event: any, ctx: any): Promise<UserNotificationMarkAsReadResponse> => {
  let success = false;

  const { id } = event.data;

  await ctx.api.gqlRequest(USER_NOTIFICATION_UPDATE_MUTATION, {
    data: {
      id,
      read: true,
    },
  });

  success = true;

  return {
    data: {
      success,
    },
  };
};
