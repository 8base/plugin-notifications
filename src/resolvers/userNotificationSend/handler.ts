import * as R from 'ramda';
import errorCodes from '@8base/error-codes';

import { CURRENT_USER__QUERY, NOTIFICATION_TEMPLATE_QUERY, NOTIFICATION_CREATE_MUTATION } from './queries';

type UserNotificationSendResponse = {
  data: {
    success: boolean;
  };
  errors?: Array<Record<string, any>>;
};

export default async (event: any, ctx: any): Promise<UserNotificationSendResponse> => {
  const { entity, templateId, templateKey, filter } = event.data;

  if (!templateKey && !templateId) {
    return {
      data: {
        success: false,
      },
      errors: [
        {
          message: 'The request is invalid.',
          code: errorCodes.ValidationErrorCode,
          details: {
            query: 'Please specify exactly one template filter parameter.',
          },
        },
      ],
    };
  }

  const currentUserPromise = ctx.api.gqlRequest(CURRENT_USER__QUERY);

  const { notificationTemplate } = await ctx.api.gqlRequest(
    NOTIFICATION_TEMPLATE_QUERY,
    {
      ...(filter ? { userFilter: filter } : {}),
      ...(templateId
        ? {
            id: templateId,
          }
        : { key: templateKey }),
    },
    {
      checkPermissions: false,
    },
  );

  const notifiers = R.pipe(
    R.pathOr([], ['roles', 'items']),
    R.map(R.pathOr([], ['users', 'items'])),
    R.flatten,
    R.uniqWith<{ id: string }, {}>((a, b) => a.id === b.id),
  )(notificationTemplate);

  const {
    user: { id: userId },
  } = await currentUserPromise;

  await ctx.api.gqlRequest(
    NOTIFICATION_CREATE_MUTATION,
    {
      data: {
        template: {
          connect: templateId
            ? {
                id: templateId,
              }
            : { key: templateKey },
        },
        actor: {
          connect: {
            id: userId,
          },
        },
        userNotifications: {
          create: notifiers.map(notifier => ({
            notifier: {
              connect: notifier,
            },
          })),
        },
        entity,
      },
    },
    {
      checkPermissions: false,
    },
  );

  return {
    data: {
      success: true,
    },
  };
};
