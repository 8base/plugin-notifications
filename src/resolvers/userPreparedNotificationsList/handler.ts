import * as R from 'ramda';
import * as Handlebars from 'handlebars';
import { generateNotificationsQuery, TableSchema } from './generateNotificationsQuery';
import { TABLES_SCHEMA_QUERY } from './queries';

type UserNotification = {
  id: string;
  read: boolean;
  createdAt: string;
  notification: {
    entity: object;
    template: {
      entityType: string;
      title: string;
      message: string;
    };
    actor: {
      id: string;
      email: string;
      firstName: string;
      lastName: string;
    };
  };
};

type UserPreparedNotification = {
  id: string;
  title: string;
  message: string;
  read: boolean;
};

type UserPreparedNotificationsListResponse = {
  data: {
    items: UserPreparedNotification[];
    count: number;
  };
};

export default async (event: any, ctx: any): Promise<UserPreparedNotificationsListResponse> => {
  const tablesSchemaResponse = await ctx.api.gqlRequest(TABLES_SCHEMA_QUERY);

  const schema: TableSchema[] | undefined = R.path(['system', 'tablesList', 'items'], tablesSchemaResponse);

  if (!schema) {
    throw new Error(`Error during get notifications entities table schema.`);
  }

  const notificationQuery = generateNotificationsQuery(schema);

  const userNotificationsListResponse = await ctx.api.gqlRequest(notificationQuery);

  const notifications: UserNotification[] = R.pathOr(
    [],
    ['userNotificationsList', 'items'],
    userNotificationsListResponse,
  );
  const count = R.pathOr(0, ['userNotificationsList', 'count'], userNotificationsListResponse);

  const userNotifications: UserPreparedNotification[] = R.map(
    ({ id, read, createdAt, notification: { actor, template, entity } }) => {
      const context = { ...entity, actor };

      const title = Handlebars.compile(template.title)(context);
      const message = Handlebars.compile(template.message)(context);

      return { title, message, id, read, createdAt };
    },
    notifications,
  );

  return {
    data: {
      items: userNotifications,
      count,
    },
  };
};
