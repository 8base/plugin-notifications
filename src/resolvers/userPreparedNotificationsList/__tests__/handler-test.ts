import handler from '../handler';
import { NOTIFICATIONS_LIST_RESPONSE, TABLES_SCHEMA_RESPONSE } from '../__fixtures__';
import { TABLES_SCHEMA_QUERY } from '../queries';

const CONTEXT = {
  api: {
    gqlRequest: jest.fn(),
  },
};

afterEach(() => {
  jest.resetAllMocks();
});

it('Should returns user prepared notifications list.', async () => {
  CONTEXT.api.gqlRequest.mockResolvedValueOnce(TABLES_SCHEMA_RESPONSE);

  CONTEXT.api.gqlRequest.mockResolvedValueOnce(NOTIFICATIONS_LIST_RESPONSE);

  const result = await handler({ data: { timezone: 'America/Los_Angeles', meta: { origin: 'test' } } }, CONTEXT);

  expect(CONTEXT.api.gqlRequest).toHaveBeenNthCalledWith(1, TABLES_SCHEMA_QUERY);

  expect(CONTEXT.api.gqlRequest.mock.calls[1][0]).toMatchSnapshot();

  expect(CONTEXT.api.gqlRequest).toHaveBeenCalledTimes(2);

  expect(result).toEqual({
    data: {
      items: [
        {
          title: 'New post "post title"',
          message: 'Vladimir Osipov created a new post "post title" 10/30/2019, 8:05 AM test',
          id: 'notficationId',
          read: false,
          createdAt: '2019-10-30T15:05:48.988Z',
          type: 'post_created',
          coverImageUrl: 'https://cdn.filestackcontent.com/security=p:policy,s:signature/fileHandle',
        },
      ],
      count: 0,
    },
  });
});
