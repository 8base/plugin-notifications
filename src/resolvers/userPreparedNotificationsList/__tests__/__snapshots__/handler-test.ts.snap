// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should returns user prepared notifications list. 1`] = `
"query UserNotifications {
  userNotificationsList {
    items {
      notification {
        template {
          entityType
          title
          message
        }
        actor {
          id
          firstName
          lastName
          email
        }
        entity {
          post {
            id
            createdAt
            updatedAt
            createdBy {
              id
              createdAt
              updatedAt
              email
              status
              origin
              is8base
              firstName
              lastName
              timezone
            }
            title
            message
            hidden
          }
          comment {
            id
            createdAt
            updatedAt
            createdBy {
              id
              createdAt
              updatedAt
              email
              status
              origin
              is8base
              firstName
              lastName
              timezone
            }
            message
            post {
              id
              createdAt
              updatedAt
              title
              message
              hidden
            }
          }
        }
      }
      id
      readed
    }
    count
  }
}
"
`;
